### PR TITLE
Feature: Add torchserve custom server bert sample

### PR DESCRIPTION
**Pull request by jagadeeshi2i**
Add torchserve custom model server with Huggingface transformers bert model

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:
```
Add torchserve custom model server with Huggingface transformers bert model
```

### CODE REVIEW

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+## Build and push the sample Docker Image
+
+The custom torchserve image is wrapped with model inside the container and serves it with KFServing.
+
+In this example we use Docker to build the torchserve image with marfile and config.properties into a container.
+
+Add the below config in config.properties to install transformers
+
+```json
+install_py_dep_per_model=true
```
[the remark this answers is not in the document]
Comment (jagadeeshi2i) [reply docs/samples/custom/torchserve/hugging-face-bert-sample.md:49]: Yes, we can and that would require rebuilding the image. Adding in marfile avoids that.

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+Modify setup_config.json
+
+```json
+{
+ "model_name":"bert-base-uncased",
+ "mode":"sequence_classification",
+ "do_lower_case":"True",
+ "num_labels":"2",
+ "save_mode":"torchscript",
```
[the remark this answers is not in the document]
Comment (jagadeeshi2i) [reply docs/samples/custom/torchserve/hugging-face-bert-sample.md:21]: Yes, I was using pretrained mode.

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+```bash
+python Download_Transformer_models.py
+```
+
+Create a requirements.txt and add `transformers` to it.
+
+Run the below command with all files in place to generate mar file.
+
+```bash
+torch-model-archiver --model-name BERTSeqClassification --version 1.0 --serialized-file Transformer_model/pytorch_model.bin --handler ./Transformer_handler_generalized.py --extra-files"Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json" -r requirements.txt
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:37]:
```suggestion
torch-model-archiver --model-name BERTSeqClassification --version 1.0 --serialized-file Transformer_model/pytorch_model.bin --handler ./Transformer_handler_generalized.py --extra-files "Transformer_model/config.json,./setup_config.json,./Seq_classification_artifacts/index_to_name.json" -r requirements.txt
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md (hunk 2 above)
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:21]:
```suggestion
 "save_mode":"pretrained",
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 4]
```diff
@@ -0,0 +1,121 @@
+# Torchserve custom server with Hugging Face bert sample
+
+## Model archive file creation
+
+[Torchserve Huggingface Transformers Bert example](https://github.com/pytorch/serve/tree/master/examples/Huggingface_Transformers)
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:5]: Can we mention that this example requires checking out this repo?

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+$inferenceservice.serving.kubeflow.org/torchserve-custom created
+```
+
+## Run a prediction
+
+The first step is to [determine the ingress IP and ports](../../../../README.md#determine-the-ingress-ip-and-ports) and set `INGRESS_HOST` and `INGRESS_PORT`
+
+```bash
+MODEL_NAME=torchserve-custom
+SERVICE_HOSTNAME=$(kubectl get route ${MODEL_NAME}-predictor-default -n <namespace> -o jsonpath='{.status.url}' | cut -d "/" -f 3)
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:90]: prefer using following to get the URL as above command won't work with canary
```
SERVICE_HOSTNAME=$(kubectl get inferenceservice ${MODEL_NAME} -n <namespace> -o jsonpath='{.status.url}' | cut -d "/" -f 3)
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md (hunk 3 above)
Comment (yuzisun) [reply docs/samples/custom/torchserve/hugging-face-bert-sample.md:37]: I have uploaded the marfile to gcs bucket, on custom yaml you can use `STORAGE_URI` env to load the model.
```
gsutil ls gs://kfserving-examples/models/torchserve/bert/
gs://kfserving-examples/models/torchserve/bert/BERTSeqClassification.mar
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+
+```bash
+# Build the container on your local machine
+
+# For CPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile -t torchserve:latest .
+
+# For GPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-argBASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve:latest .
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:73]:
```suggestion
DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve:latest .
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+## Run a prediction
+
+The first step is to [determine the ingress IP and ports](../../../../README.md#determine-the-ingress-ip-and-ports) and set `INGRESS_HOST` and `INGRESS_PORT`
+
+```bash
+MODEL_NAME=torchserve-custom
+SERVICE_HOSTNAME=$(kubectl get inferenceservice ${MODEL_NAME} -n <namespace> -o jsonpath='{.status.url}' | cut -d "/" -f 3)
+
+curl -v -H "Host: ${SERVICE_HOSTNAM}" http://${INGRESS_HOST}:${INGRESS_PORT}/predictions/BERTSeqClassification -T serve/examples/Huggingface_Transformers/sample_text.txt
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:104]:
```suggestion
curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/predictions/BERTSeqClassification -T serve/examples/Huggingface_Transformers/sample_text.txt
```

File: docs/samples/custom/torchserve/bert.yaml [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+kind: "InferenceService"
+metadata:
+  name: torchserve-custom
+spec:
+  predictor:
+    containers:
+    - image: {username}/torchserve_custom:latest
+      name: kfserving-container
+      ports:
+        - containerPort: 8080
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/bert.yaml:11]:
```suggestion
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+## Create the InferenceService
+
+In the `torchserve-custom.yaml` file edit the container image and replace {username} with your Docker Hub username.
+
+Apply the CRD
+
+```bash
+kubectl apply -f torchserve-custom.yaml
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:87]:
```suggestion
kubectl apply -f bert.yaml
```

File: docs/samples/custom/torchserve/bert.yaml [hunk 10]
```diff
@@ -0,0 +1,12 @@
+apiVersion: "serving.kubeflow.org/v1beta1"
+kind: "InferenceService"
+metadata:
+  name: torchserve-custom
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/bert.yaml:4]:
```suggestion
  name: torchserve-bert
```

File: docs/samples/custom/torchserve/bert.yaml [hunk 11]
```diff
@@ -0,0 +1,12 @@
...
+spec:
+  predictor:
+    containers:
+    - image: {username}/torchserve_custom:latest
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/bert.yaml:8]:
```suggestion
    - image: {username}/torchserve-bert:latest
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+To build and push with Docker Hub, run these commands replacing {username} with your Docker Hub username:
+
+[Dockerfile for torchserve image building](https://github.com/pytorch/serve/blob/master/docker/Dockerfile)
+
+
+```bash
+# Build the container on your local machine
+
+# For CPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile -t torchserve:latest .
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:70]:
```suggestion
DOCKER_BUILDKIT=1 docker build --file Dockerfile -t torchserve-bert:latest .
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+# For GPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve:latest .
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:73]:
```suggestion
DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve-bert-gpu:latest .
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+Apply the CRD
+
+```bash
+kubectl apply -f torchserve-custom.yaml
+```
+
+Expected Output
+
+```bash
+$inferenceservice.serving.kubeflow.org/torchserve-custom created
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:93]:
```suggestion
$inferenceservice.serving.kubeflow.org/torchserve-bert created
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+```
+
+## Run a prediction
+
+The first step is to [determine the ingress IP and ports](../../../../README.md#determine-the-ingress-ip-and-ports) and set `INGRESS_HOST` and `INGRESS_PORT`
+
+```bash
+MODEL_NAME=torchserve-custom
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:101]:
```suggestion
MODEL_NAME=torchserve-bert
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+curl -v -H "Host: ${SERVICE_HOSTNAME}" http://${INGRESS_HOST}:${INGRESS_PORT}/predictions/BERTSeqClassification -T serve/examples/Huggingface_Transformers/sample_text.txt
+```
+
+Expected Output
+
+```bash
+*   Trying 44.239.20.204...
+* Connected to a881f5a8c676a41edbccdb0a394a80d6-2069247558.us-west-2.elb.amazonaws.com (44.239.20.204) port 80 (#0)
+> PUT /predictions/BERTSeqClassification HTTP/1.1
+> Host: torchserve-custom-predictor-default.kfserving-test.example.com
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:113]:
```suggestion
> Host: torchserve-bert.kfserving-test.example.com
```

File: docs/samples/custom/torchserve/hugging-face-bert-sample.md [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+< cache-control: no-cache; no-store, must-revalidate, private
+< content-length: 8
+< date: Wed, 04 Nov 2020 10:54:49 GMT
+< expires: Thu, 01 Jan 1970 00:00:00 UTC
+< pragma: no-cache
+< x-request-id: 4b54d3ac-185f-444c-b344-b8a785fdeb50
+< x-envoy-upstream-service-time: 2085
+< server: istio-envoy
+<
+* Connection #0 to host torchserve-custom-predictor-default.kfserving-test.example.com left intact
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/hugging-face-bert-sample.md:131]:
```suggestion
* Connection #0 to host torchserve-bert.kfserving-test.example.com left intact
```

**Review by yuzisun — COMMENTED** [review]
Can we move this example to a sub directory under torchserve?

File: docs/samples/custom/torchserve/bert-sample/hugging-face-bert-sample.md [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+# Build the container on your local machine
+
+# For CPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile -t torchserve-bert:latest .
+
+# For GPU
+DOCKER_BUILDKIT=1 docker build --file Dockerfile --build-arg BASE_IMAGE=nvidia/cuda:10.1-cudnn7-runtime-ubuntu18.04 -t torchserve-bert-gpu:latest .
+
+# Push the container to docker registry
+docker tag torchserve:latest {username}/torchserve:latest
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/bert-sample/hugging-face-bert-sample.md:76]:
```suggestion
docker tag torchserve-bert:latest {username}/torchserve-bert:latest
```

File: docs/samples/custom/torchserve/bert-sample/hugging-face-bert-sample.md [hunk 19]
```diff
@@ -0,0 +1,133 @@
...
+docker push {username}/torchserve:latest
```
Comment (yuzisun) [inline docs/samples/custom/torchserve/bert-sample/hugging-face-bert-sample.md:77]:
```suggestion
docker push {username}/torchserve-bert:latest
```

**Review by yuzisun — COMMENTED** [review]
Great example! thanks @jagadeeshi2i !

/lgtm
/approve

**Pull request merged**